Initial cortex-m0+ port (mostly stm32g0)

cbiffle opened the pull request:
This builds on @jperkin's port and my notes in #395, applying some optimizations and fixes to support more and smaller targets, and rebasing it to take our recent system changes into account.

~~Note that this is currently contingent on the `static-tcbs` PR #393. This is because the changes in that PR make it easier to conserve kernel memory, which is important on devices with down to 8 kiB of RAM.~~ #393 was merged a while back.

Outstanding issues:
- [x] figure out why Humility doesn't work reliably
- [x] change fault mapping to not say "Illegal Instruction"
- [x] don't check in anything that uses semihosting by default because it makes images hard to test
- [x] test DebugBI support like Humility-Hiffy
- [x] get test suite working

mkeeter commented on drv/stm32g0-rcc-api/src/lib.rs:
```diff
@@ -0,0 +1,185 @@
+// - RCC_APBENR1[31:0] and RCC_APBRSTR1[31:0] are indices 95-64.
+// - RCC_APBENR2[31:0] and RCC_APBRSTR2[31:0] are indices 127-96.
+//
+macro_rules! iop {
```
What about using a `const fn` here instead of a macro?

(it looks like this is also applicable to the `stm32h7-rcc-api`!)

cbiffle replied:
I believe that, at the time the original code (which was for the f4) was written, const fns were not permitted in this syntactic position (the definition of enum values)? But they definitely are now! Will fix this one. Would you be willing to file a brief issue to ensure that we go get the f4/h7/etc?

mkeeter replied:
Done, opened #404 [PR NOT FOUND]

mkeeter commented on drv/stm32g0-rcc/src/main.rs:
```diff
@@ -0,0 +1,213 @@
+
+    // Field messages.
+    // Ensure our buffer is aligned properly for a u32 by declaring it as one.
+    let mut buffer = [0u8; idl::INCOMING_SIZE];
```
...this doesn't look like a `u32`, contrary to the comment above

(same issue exists in the H7 RCC task)

cbiffle replied:
That commit is obsolete -- I should have removed it when I did the stm32h7 rcc Idol conversion, and it looks like it got copied into the stm32g0 version. Will fix.

mkeeter commented on lib/armv6m-atomic-hack/src/lib.rs:
```diff
@@ -0,0 +1,94 @@
+
+#![no_std]
+
+use core::sync::atomic::{AtomicBool, AtomicU32, Ordering};
```
You probably want `#[cfg(armv6m)]` here too so that Github Actions doesn't get _really mad_ (or wrap the whole file in that conditional?)

cbiffle replied:
Ha, good point. Wrapping the whole file in the conditional made things unhappy, iirc.

mkeeter commented on lib/armv6m-atomic-hack/src/lib.rs:
```diff
@@ -0,0 +1,94 @@
+        Ordering::SeqCst => (o, o),
+        Ordering::Acquire => (Ordering::Acquire, Ordering::Relaxed),
+        Ordering::Release => (Ordering::Relaxed, Ordering::Release),
+        _ => panic!(),
```
It looks like the `match` covers all options – why do we need a default case here?

cbiffle replied:
Because of the `#[non_exhaustive]` attribute on https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html - that's the authors reserving the right to add new variants in the future.

mkeeter replied:
Aha, makes sense!

mkeeter commented on sys/kern/src/arch/arm_m.rs:
```diff
@@ -1151,6 +1457,54 @@ bitflags::bitflags! {
+    // We are now going to force a fault on our current task and directly
+    // switch to a task to run.  (It may be tempting to use PendSV here,
+    // but that won't work on ARMv8-M in the presence of MPU faults on
+    // PSP:  even with PendSV pending, ARMv8-M will generate a MUNSTKERR
```
This part of the comment isn't relevant in an explicitly `armv6` block

mkeeter commented on sys/userlib/src/macros.rs:
```diff
@@ -33,7 +33,18 @@ macro_rules! sys_log {
+    ($s:expr, $($tt:tt)*) => {};
+}
+
+#[cfg(not(any(
```
YMMV, but whenever I see both positive and negative `#[cfg...` branches, I prefer to replace it with a `cfg_if` block (and then use `else` for the negative case).

cbiffle replied:
Heh, amusingly, I did this for a _lot_ of cases in this changeset, but not this one. Fixing.

cbiffle commented on task/pong/Cargo.toml:
```diff
@@ -3,11 +3,12 @@ name = "task-pong"
 
-# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+[features]
+panic-messages = ["userlib/panic-messages"]
```
It does change that, yes. We could edit all callers; I haven't gone ahead and done that because... well, okay, the most proximal reason is likely laziness, but my thought process was

- pong doesn't really panic.
- pong is a dev demo that isn't in production.

I can be talked out of this.

mkeeter replied:
Alright, I don't feel strongly about it!

cbiffle replied:
Frankly we kind of need better app-level controls for this sort of thing, so that making something like this configurable doesn't require running around adding a bazillion `features = ["panic-messages"]` to every app.toml. I'm not sure what that looks like yet though. Very open to suggestions.

cbiffle commented:
FWIW as of last night, most of Humility (including Hiffy-driven commands like gpio) is working on m0+. I think we're just waiting on stack traces, which is in progress.

cbiffle commented:
- Rebased atop recent master
- Got the kernel test suite passing (which required adding semihosting support to it)
- Fixed one asm mistake that revealed, in the `reply_fault` syscall stub.

cbiffle commented:
- Updated STM32G031 port to actually work on the Nucleo G031 board.
- Fixed an xtask bug along the way
- Noticed that the cortex-m0 port was acting like exception frames contained floating point registers; fixed this for a size reduction.

cbiffle commented:
Okay, you can now replicate my hiffy demo from Friday using this branch if you're so inclined. I'm hoping to get this merged soon.
